alerts/alerts.py

```diff
@@ -0,0 +1,515 @@
+
+    if top_artist != historic_top_artist:
+        logging.info("Top artist has changed: Alerting subscribers")
+        for email in sub_emails:
+            send_top_artist_alert(email, top_artist)
+        logging.info("All top artist change alerts sent successfully")
+
+    else:
+        logging.info("No change detected in top artist")
+
+
+def alert_top_genre_change(cursor: RealDictCursor) -> None:
+    """
+    Checks if the historical top genre has recently changed,
+    and sends an alert to all subscribers if it has.
+    """
+    top_genre = query_top_genre(cursor)
+    historic_top_genre = query_historic_top_genre(cursor)
+    sub_emails = get_general_subscriber_emails(cursor)
+
+    logging.info("Checking for change in top genre")
+
+    if top_genre != historic_top_genre:
+        logging.info("Top genre has changed: Alerting subscribers")
+        for email in sub_emails:
+            send_top_genre_alert(email, top_genre)
+        logging.info("All top genre change alerts sent successfully")
+
+    else:
+        logging.info("No change detected in top genre")
+
+
+def calculate_genre_sales_delta(cursor, genre) -> float:
+    """
+    Calculates the percentage difference in sales numbers for a given genre
+    within the last alert interval compared to the comparison period.
+    """
+    recent_sales = query_genre_sales(cursor, genre)
+    historic_sales = query_historic_genre_sales(cursor, genre)
+
+    sales_delta = 100 * (recent_sales - historic_sales) / historic_sales
+
+    return sales_delta
+
+
+def alert_subscribed_genres(cursor: RealDictCursor) -> None:
+    """
+    Checks each subscribed genre for if it has grown in popularity
+    past the notification threshold recently, and sends alerts
+    for each one that has to those who are subscribed.
+    """
+    logging.info(
+        "Checking for increases in subscribed genre sales above the alert threshold")
+
+    genres = get_subscribed_genres(cursor)
+    for genre in genres:
+        sales_delta = calculate_genre_sales_delta(cursor, genre)
+        if sales_delta > GENRE_NOTIFICATION_THRESHOLD:
+            logging.info(
+                "Alert threshold surpassed for '%s': Alerting subscribers", genre)
+            emails = get_genre_subscriber_emails(cursor, genre)
+            for email in emails:
+                send_chosen_genre_alert(email, genre, sales_delta)
+            logging.info("All '%s' alerts sent successfully", genre)
+        else:
+            logging.info(
+                "'%s' sales did not surpass the alert threshold", genre)
+
+
+def main() -> None:
+    """The main function that generates and sends all alerts"""
+    load_dotenv()
+    config_log()
+    db_conn = get_connection()
+    db_cursor = get_cursor(db_conn)
+    alert_top_artist_change(db_cursor)
+    alert_top_genre_change(db_cursor)
+    alert_subscribed_genres(db_cursor)
+    db_conn.close()
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Really nice clean code - I cannot fault it. My only suggestion is the script is 515 lines - quite long! Maybe it would be better to split into a query script and an alerts script?

Reply:
Yeah I agree with Ellie here, it's amazing, but very very long!!